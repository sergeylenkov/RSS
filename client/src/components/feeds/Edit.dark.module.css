.icon {
	composes: icon from './Edit.module.css';
	stroke: #a5a5a5;
	fill: #a5a5a5;
}