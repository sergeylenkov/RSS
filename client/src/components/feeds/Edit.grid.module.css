.container {
  margin: auto 0 auto auto;
}