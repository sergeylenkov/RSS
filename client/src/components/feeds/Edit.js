import React from 'react';
import { Icon, Icons } from '../Icon.js';

import styles from './Edit.module.css';

export class FeedEdit extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isFormVisible: false
        }

        this.linkFieldRef = React.createRef();
    }

    render() {
        let className = styles.container;

        if (this.props.isDisabled) {
            className += ` ${styles.disabled}`;
        }

        if (this.props.isScrolled) {
            className += ` ${styles.scrolled}`;
        }

        if (this.state.isFormVisible) {
            return (
                <div className={className}>
                    <div className={styles.form}>
                        <input className={styles.linkField} placeholder="ссылка на канал" type="text" ref={this.linkFieldRef} />
                        <button className={styles.formButton} onClick={() => this.onAdd()}><div className={styles.icon}><Icon svg={Icons.add}/></div></button>
                        <button className={styles.formButton} onClick={() => this.onCloseForm()}><div className={styles.icon}><Icon svg={Icons.close}/></div></button>
                    </div>
                </div>
            );
        } else {
            let iconClassName = styles.icon;

            if (this.props.isEdited) {
                iconClassName += ` ${styles.edited}`;
            }

            return (
                <div className={className}>
                    <button className={styles.button} onClick={() => this.onShowForm()}><div className={styles.icon}><Icon svg={Icons.add}/></div></button>
                    <button className={styles.button} onClick={() => this.props.onEdit()}><div className={iconClassName}><Icon svg={Icons.edit}/></div></button>
                </div>
            );
        }
    }

    onAdd() {        
        const value = this.linkFieldRef.current.value;

        if (value.length > 0) {
            this.props.onAdd(value);
        }
    }

    onShowForm() {
        this.setState({
            isFormVisible: true
        });
    }

    onCloseForm() {
        this.setState({
            isFormVisible: false
        });
    }
}