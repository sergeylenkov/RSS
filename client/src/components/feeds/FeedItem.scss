.feed-item {
  display: flex;
  flex-direction: row;
  width: 100%;
  height: 30px;

  &__button {
    display: flex;
    flex-direction: row;
    overflow: hidden;
    cursor: pointer;
  }

  &__icon {
    flex-shrink: 0;
    width: 16px;
    height: 16px;
    margin: auto 10px auto 0;
    background-image: "";
    background-repeat: no-repeat;
    background-size: cover;
    background-color: #ebebeb;
  }

  &__label {
    margin: auto 0;
    font-size: 13px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    color: #000000;

    &_selected {
      font-weight: bold;
    }
  }

  &__counter {
    margin: auto 0 auto 10px;
    padding: 2px 5px;
    font-size: 11px;
    color: #ffffff;
    border-radius: 2px;
    background-color: #4da0da;
  }

  &__edit_panel {
    display: flex;
    flex-direction: row;
    width: 100%;
  }

  &__edit_button {
    padding: 0;
    margin: auto 0 auto 10px;
    display: flex;
    cursor: pointer;

    .icon {
      width: 16px;
      height: 16px;
      stroke: #000000;
      stroke-width: 2px;
      fill: #000000;

      svg {
        width: 100%;
        height: 100%;
      }
    }
  }

  &__title_field {
    margin: auto 10px auto 0;
    font-size: 13px;
    flex-grow: 1;
  }
}

/*
* grid
*/

.grid .feed-item {
  width: auto;
  height: 30px;
  margin: auto 20px auto 0;
}

/*
* theme
*/

.dark .feed-item {
  &__icon {
    background-color: #bfbfbf;
  }

  &__label {
    color: #bfbfbf;

    &--selected {
      color: #f3f3f3;
    }
  }

  &__edit_button {
    .icon {
      stroke: #a5a5a5;
      fill: #a5a5a5;
    }
  }
}

@media (max-width: 768px) {
  .feed-item {
    flex-shrink: 0;
    width: auto;
    margin-right: 20px;
  }
}
