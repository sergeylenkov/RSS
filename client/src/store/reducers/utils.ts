import { Entry, Feed } from '../../data';
import { State } from './index';

const feedsDict: { [key: string]: Feed } = {};

export function updateFeeds(state: State, feeds: Feed[]): State {
  feeds.forEach((feed: Feed) => {
    feedsDict[feed.id] = feed;

    const iconId = `icon_${feed.id}`;

    feed.icon = localStorage.getItem(iconId) || feed.iconUrl;

    if (!localStorage.getItem(iconId)) {
      toDataURL(feed.iconUrl).then(data => {
        localStorage.setItem(iconId, data);
      });    
    }
  });

  return {
    ...state,
    isUpdating: false,
    isInitialized: true,
    feeds: [...feeds],
  };
}

export function updateFeed(state: State, id: number, newFeed: Feed): State {
  const feeds = state.feeds.map((feed: Feed) => {
    return feed.id === id ? { ...feed, ...newFeed } : feed;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

export function addFeed(state: State, feed: Feed): State {
  const feeds = [...state.feeds];

  feeds.push(feed);

  return {
    ...state,
    feeds: feeds,
  };
}

export function removeFeed(state: State, id: number): State {
  const feeds = state.feeds.filter((feed: Feed) => {
    return feed.id !== id;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

export function selectFeed(state: State, id: number): State {
  const selectedFeeds = [...state.selectedFeeds];

  const index = selectedFeeds.indexOf(id);

  if (index === -1) {
    selectedFeeds.push(id);
  } else {
    selectedFeeds.splice(index, 1);
  }

  const entries = filterEntries(selectedFeeds, state.allEntries);

  return {
    ...state,
    selectedFeeds: selectedFeeds,
    entries: [...entries],
  };
}

export function updateUnviewedCount(state: State): State {
  const feeds = [...state.feeds];
  let totalCount = 0;

  feeds.forEach((feed: Feed) => {
    const count = state.entries.filter((entry: Entry) => {
      return entry.feedId === feed.id && !entry.isViewed;
    }).length;

    feed.count = count;
    totalCount = totalCount + count;
  });

  return {
    ...state,
    feeds: feeds,
    unviewedCount: totalCount,
  };
}

export function updateViewed(state: State, ids: number[]): State {
  let entries = state.entries.map((entry: Entry) => {
    if (ids.indexOf(entry.id) !== -1) {
      return { ...entry, isViewed: true };
    }

    return entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateFavorite(
  state: State,
  id: number,
  isFavorite: boolean
): State {
  let entries = state.entries.map((entry: Entry) => {
    return entry.id === id ? { ...entry, isFavorite: isFavorite } : entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateRead(state: State, id: number, isRead: boolean): State {
  let entries = state.entries.map((entry: Entry) => {
    return entry.id === id ? { ...entry, isRead: isRead } : entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateEntries(state: State, entries: Entry[]): State {
  updateFeedsInEntries(entries);

  const filteredEntries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    isUpdating: false,
    entriesCount: entries.length,
    entries: [...filteredEntries],
    allEntries: [...entries],
  };
}

export function updateEntriesCount(state: State): State {
  const feeds = [...state.feeds];

  feeds.forEach((feed: Feed) => {
    feed.count = state.entries.filter((entry: Entry) => {
      return entry.feedId === feed.id;
    }).length;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

function filterEntries(selectedFeeds: number[], allEntries: Entry[]): Entry[] {
  let entries: Entry[] = [];

  if (selectedFeeds.length > 0) {
    entries = allEntries.filter((entry: Entry) => {
      return selectedFeeds.includes(entry.feedId);
    });
  } else {
    entries = allEntries;
  }

  return entries;
}

function getFeedById(id: number): Feed {
  return feedsDict[id];
}

function updateFeedsInEntries(entries: Entry[]): void {
  entries.forEach((entry) => {
    entry.feed = getFeedById(entry.feedId);
  });
}

function toDataURL(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fetch(url)
    .then(response => response.blob())
    .then(blob => {
      const reader = new FileReader()

      reader.onloadend = () => resolve(reader.result as string)
      reader.onerror = reject
      reader.readAsDataURL(blob)
    });
  });  
}