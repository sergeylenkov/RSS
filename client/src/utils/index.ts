export const debounce = <F extends (...args: unknown[]) => unknown>(
  func: F,
  waitFor: number
) => {
  let timeout = 0;

  const debounced = (...args: unknown[]) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => func(...args), waitFor) as unknown as number;
  };

  return debounced as unknown as (...args: Parameters<F>) => ReturnType<F>;
};

export function isSystemDarkTheme(): boolean {
  return (
    window.matchMedia &&
    window.matchMedia('(prefers-color-scheme: dark)').matches
  );
}
