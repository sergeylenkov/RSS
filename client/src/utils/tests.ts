export const action = (): void => { return };