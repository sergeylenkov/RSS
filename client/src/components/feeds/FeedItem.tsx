import React, { useRef } from 'react';
import { Bem } from '../../utils/bem';
import { CheckmarkIcon, TrashIcon } from '../Icons';

import './FeedItem.scss';

const block = new Bem('feed-item');

interface FeedProps {
  id: number;
  icon: string;
  title: string;
  count: number;
  isEditing: boolean;
  isSelected: boolean;
  onSelect: (id: number) => void;
  onDelete: (id: number) => void;
  onChange: (id: number, title: string) => void;
}

function FeedItem({
  id,
  icon,
  title,
  count,
  isEditing,
  isSelected,
  onSelect,
  onDelete,
  onChange,
}: FeedProps): JSX.Element {
  const titleFieldRef = useRef<HTMLInputElement>(null);

  const onChangeTitle = (): void => {
    if (titleFieldRef && titleFieldRef.current) {
      onChange(id, titleFieldRef.current.value.trim());
    }
  };

  const labelClass = block
    .getElement('label')
    .addModifier(isSelected ? 'selected' : '')
    .toString();
  const counterClass = block.getElement('counter').toString();
  const iconClass = block.getElement('icon').toString();

  return (
    <div className={block.toString()}>
      {isEditing ? (
        <div className="feed-item__edit_panel">
          <div
            className={iconClass}
            style={{ backgroundImage: `url(${icon})` }}
          />
          <input
            ref={titleFieldRef}
            className="feed-item__title_field"
            defaultValue={title}
            type="text"
          />
          <button className="feed-item__edit_button" onClick={onChangeTitle}>
            <div className="icon">
              <CheckmarkIcon />
            </div>
          </button>
          <button
            className="feed-item__edit_button"
            onClick={() => onDelete(id)}
          >
            <div className="icon">
              <TrashIcon />
            </div>
          </button>
        </div>
      ) : (
        <button
          className={block.getElement('button').toString()}
          onClick={() => onSelect(id)}
        >
          <div
            className={iconClass}
            style={{ backgroundImage: `url(${icon})` }}
          />
          <div className={labelClass}>{title}</div>
          <div className={counterClass}>{count}</div>
        </button>
      )}
    </div>
  );
}

export default FeedItem;
