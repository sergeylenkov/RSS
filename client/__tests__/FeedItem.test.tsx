import React from 'react';
import renderer from 'react-test-renderer';
import { FeedItem } from '../src/components/feeds/FeedItem';
import { action } from '../src/utils/tests';

test('FeedItem isSelected = true', () => {
  const render = renderer.create(
    <FeedItem
      id={1}
      icon={''}
      title={'Test'}
      count={0}
      isEditing={false}
      isSelected
      onSelect={action}
      onChange={action}
      onDelete={action}
    />
  );
  expect(render).toMatchSnapshot();
});

test('FeedItem isEditing = true', () => {
  const render = renderer.create(
    <FeedItem
      id={1}
      icon={''}
      title={'Test'}
      count={0}
      isEditing
      isSelected={false}
      onSelect={action}
      onChange={action}
      onDelete={action}
    />
  );
  expect(render).toMatchSnapshot();
});
