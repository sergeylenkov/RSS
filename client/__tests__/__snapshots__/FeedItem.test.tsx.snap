// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedItem correct rendering 1`] = `
<div
  className="feed-item"
>
  <button
    className="feed-item__button"
    onClick={[Function]}
  >
    <div
      className="feed-item__icon"
      style={
        Object {
          "backgroundImage": "url()",
        }
      }
    />
    <div
      className="feed-item__label feed-item__label_selected"
    >
      Test
    </div>
    <div
      className="feed-item__counter"
    >
      0
    </div>
  </button>
</div>
`;
